File: python/키 순서(2458).py
# 나한테 오는 노드(나보다 작음) + 내가 갈 수 있는 노드(나보다 큼) = n-1 -> 키가 몇 번째인지 알 수 있음
# 플로이드 와샬 알고리즘

import sys
input = sys.stdin.readline

n, m = tuple(map(int, input().split()))
graph = [[0]*(n+1) for _ in range(n+1)]
for _ in range(m):
    x, y = tuple(map(int, input().split()))
    graph[x][y] = 1

for k in range(1, n+1):
    for i in range(1, n+1):
        for j in range(1, n+1):
            if graph[i][k] and graph[k][j]:
                graph[i][j] = 1

answer = 0
for i in range(1, n+1):
    bigger = sum(graph[i][1:])
    smaller = sum(graph[j][i] for j in range(1, n+1))
    if bigger + smaller == n-1:
        answer += 1

print(answer)

File: js/키 순서(2458).js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const graph = Array.from({length: n+1}, () => Array(n+1).fill(0));
for (let i=1; i<=m; i++) {
    let [x,y] = input[i].split(' ').map(Number);
    graph[x][y] = 1;
}

// 플로이드 와샬 알고리즘
// i < k, k < j 이면 i < j 
for (let k=1; k<n+1; k++) {
    for (let i=1; i<n+1; i++) {
        for (let j=1; j<n+1; j++) {
            if (graph[i][k] && graph[k][j]) {
                graph[i][j] = 1;
            }
        }
    }
}

let answer = 0;
for (let i=1; i<n+1; i++) {
    let bigger = graph[i].slice(1).reduce((acc, cur) => acc + cur, 0);
    let smaller = 0;
    for (j=1; j<n+1; j++) {
        smaller += graph[j][i];
    }
    if (bigger+smaller === n-1) {
        answer++;
    }
}

console.log(answer);